### PR TITLE
EZP-30350: in-memory breaks BC for cache clearing

### PR DESCRIPTION
**Pull request by andrerom**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30350](https://jira.ez.no/browse/EZP-XXXXX)
| **Bug**| yes
| **Target version** | `7.5`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

Fixes:
1. BC break, so that when Developer or handler code calls `$pool->clear()` or `$pool->invalidateTags(['language-2']);`, it will also clear in-memory cache
2. Removes need for abstracting cache clearing this in handler, Reverts changes for this in handlers that has gotten in-memory support already
3. Un-blocks possibility to introduce cache for content-info and locations, as we need to make sure any call to `$pool->invalidateTags(['content-2']);` still works as intended, which is the case in almost all remaining handlers

It also adjusts constructor of AbstractInMemoryHandler to make it possible to reuse from external handlers _(TagsBundle, and others)_ without having to have knowledge of the internal classes used for cache.

This is needed for:
- Adding in-memory cache for TagsBundle
- _(Post release)_ Add cache for ContentInfo & Locations 

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/Persistence/Cache/Adapter/InMemoryClearingProxyAdapter.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ */
+class InMemoryClearingProxyAdapter implements TagAwareAdapterInterface
+{
+    /**
+     * @var TagAwareAdapterInterface
+     */
+    private $pool;
+
+    /**
+     * @var InMemoryCache
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Cache/Adapter/InMemoryClearingProxyAdapter.php]: this should use FQCN, same with adapter above

File: eZ/Publish/Core/Persistence/Cache/Adapter/InMemoryClearingProxyAdapter.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+     */
+    private $inMemory;
+
+    public function __construct(TagAwareAdapterInterface $pool, InMemoryCache $inMemory)
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Cache/Adapter/InMemoryClearingProxyAdapter.php]: Maybe add `{@inheritDoc}` on the methods here at least

File: eZ/Publish/Core/Persistence/Cache/Tests/AbstractBaseHandlerTest.php [hunk 3]
```diff
@@ -36,7 +36,7 @@
 abstract class AbstractBaseHandlerTest extends TestCase
 {
     /**
-     * @var \Symfony\Component\Cache\Adapter\TagAwareAdapterInterface|\PHPUnit\Framework\MockObject\MockObject
+     * @var \eZ\Publish\Core\Persistence\Cache\Adapter\InMemoryClearingProxyAdapter|\PHPUnit\Framework\MockObject\MockObject
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Cache/Tests/AbstractBaseHandlerTest.php]: maybe no need to change this, as signature is still the same

File: eZ/Publish/Core/Persistence/Cache/AbstractInMemoryHandler.php [hunk 4]
```diff
@@ -43,31 +42,43 @@ abstract class AbstractInMemoryHandler
...
-        $this->inMemory = $inMemory;
-
-        $this->init();
+        $this->setCacheDependencies($cache, $logger, $inMemory);
     }
 
     /**
-     * Optional dunction to initialize handler without having to overload __construct().
+     * @api May be used from handlers that needs to inject something else then PersistenceHandler to __construct(), to
+     *      avoid having to have intimate knowledge of internal classes used by AbstractInMemoryHandler in signature.
```
Comment (adamwojs) [inline eZ/Publish/Core/Persistence/Cache/AbstractInMemoryHandler.php]: Hiding dependencies using argument unpacking not a good approach. If `AbstractInMemoryHandler` requires any internal dependency in the constructor then it should be internal as well.

Comment (andrerom) [reply eZ/Publish/Core/Persistence/Cache/AbstractInMemoryHandler.php]: @adamwojs So should we?:
- keep init()
- mark constructor as internal
- anything else?

Comment (andrerom) [reply eZ/Publish/Core/Persistence/Cache/AbstractInMemoryHandler.php]: > should be internal as well.

or what did you means here actually?

File: eZ/Publish/Core/Persistence/Cache/Tests/Adapter/InMemoryClearingProxyAdapterTest.php [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+            ['clear', null],
+        ];
+    }
+
+    /**
+     * @param $key
+     * @param null $value If null the cache item will be assumed to be a cache miss here.
+     * @param int $defaultLifetime
+     *
+     * @return CacheItem
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Cache/Tests/Adapter/InMemoryClearingProxyAdapterTest.php]: Nitpick: FQCN

**Pull request merged**